core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java

```diff
@@ -0,0 +1,572 @@
+        partitionAssignments = new IntArrayList[partitionCount];
+        for (int i = 0; i < partitionAssignments.length; i++) {
+            partitionAssignments[i] = new IntArrayList();
+            partitionPageSinks[i] = null;
+        }
+
+        this.tableWriterInfoSupplier = createTableWriterInfoSupplier(pageSinkPeakMemoryUsage, statisticsTiming, partitionPageSinks);
+        this.operatorContext.setInfoSupplier(tableWriterInfoSupplier);
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public void finish()
+    {
+        OperationTimer timer = new OperationTimer(statisticsCpuTimerEnabled);
+        statisticAggregationOperator.finish();
+        timer.end(statisticsTiming);
+
+        ImmutableList.Builder<ListenableFuture<?>> blockedFutures = ImmutableList.builder();
+        blockedFutures.add(blocked);
+        blockedFutures.add(statisticAggregationOperator.isBlocked());
+
+        if (state == State.RUNNING) {
+            state = State.FINISHING;
+            ImmutableList.Builder<CompletableFuture<Collection<Slice>>> futures = ImmutableList.builder();
+            for (ConnectorPageSink pageSink : partitionPageSinks) {
+                if (pageSink != null) {
+                    CompletableFuture<Collection<Slice>> finish = pageSink.finish();
+                    futures.add(finish);
+                    blockedFutures.add(toListenableFuture(finish));
+                }
+            }
+            pageSinkFinishFutures = futures.build();
+            updateWrittenBytes();
+        }
+        this.blocked = asVoid(allAsList(blockedFutures.build()));
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state == State.FINISHED && blocked.isDone();
+    }
+
+    @Override
+    public ListenableFuture<Void> isBlocked()
+    {
+        return blocked;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        if (state != State.RUNNING || !blocked.isDone()) {
+            return false;
+        }
+        return statisticAggregationOperator.needsInput();
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+        checkState(needsInput(), "Operator does not need input");
+
+        OperationTimer timer = new OperationTimer(statisticsCpuTimerEnabled);
+        statisticAggregationOperator.addInput(page);
+        timer.end(statisticsTiming);
+        ListenableFuture<Void> blockedOnAggregation = statisticAggregationOperator.isBlocked();
+
+        ImmutableList.Builder<ListenableFuture<?>> blockedOnWrites = ImmutableList.builder();
+        blockedOnWrites.add(blockedOnAggregation);
+
+        Page partitionedPage = partitionPagePreparer.apply(page);
+
+        for (int position = 0; position < partitionedPage.getPositionCount(); position++) {
+            int partition = partitionFunction.getPartition(partitionedPage, position);
```

Comment:
Partitioning is required here because it is possible that `ScaleWriterPartitioningExchanger` sends out pages which contain multiple partitions. Basically, `ScaleWriterPartitioningExchanger` assign multiple set of partitions to a single writer after considering scaling stuff, and then create a page which contains all these partitions.

Another approach could be to create a page per partition which include the partition unique id as a separate RLE channel. However, it increases the complexity since you need to describe this channel as part of plan and there can be very small pages which can impact performance. Therefore we've decided not to go with this soln.

In future, we could pre-generate the hashes to save the computation. https://github.com/trinodb/trino/issues/14184

Reply:
I'll add more comments to this class

Reply:
> it is possible that ScaleWriterPartitioningExchanger sends out pages which contain multiple partitions

Ok, got it. 
But when I look at `HivePageSink` or `IcebergPageSink` I see it partitions the page anyway because it has to split the data into files so why partitioning here + multiple `ConnectorPageSink`s is needed? Should we noty leave it to the `ConnectorPageSink`?

Reply:
We need partitioning here along with multiple `ConnectorPageSinks` to collect statistics around partition level physicaWrittenBytes which is essential for scaling. In the ideal world, we could get rid of partitioning from pageSink if we do partitioning directly inside the local exchange itself using pageIndexer which handles hash collisions also. However, it again adds complexity which I mentioned in the above comment.